File: backend/app/api/vehicles.py
from fastapi import APIRouter, Depends, status

from app.core.deps import get_current_user, require_admin
from app.models.vehicle import Vehicle
from app.schemas.vehicle import VehicleCreateRequest, VehicleResponse, VehicleUpdateRequest
from app.services.vehicle_service import (
    create_vehicle,
    delete_vehicle,
    list_vehicles,
    search_vehicles,
    update_vehicle,
)

router = APIRouter(prefix="/api/vehicles", tags=["vehicles"])


def _to_response(vehicle: Vehicle) -> VehicleResponse:
    return VehicleResponse(
        id=str(vehicle.id),
        make=vehicle.make,
        model=vehicle.model,
        category=vehicle.category,
        price=vehicle.price,
        quantity=vehicle.quantity,
    )


@router.post("", response_model=VehicleResponse, status_code=status.HTTP_201_CREATED)
async def create(data: VehicleCreateRequest, _=Depends(get_current_user)):
    vehicle = await create_vehicle(data)
    return _to_response(vehicle)

@router.get("/search", response_model=list[VehicleResponse])
async def search(
    make: str | None = None,
    model: str | None = None,
    category: str | None = None,
    min_price: float | None = None,
    max_price: float | None = None,
    _=Depends(get_current_user),
):
    vehicles = await search_vehicles(make, model, category, min_price, max_price)
    return [_to_response(v) for v in vehicles]


@router.get("", response_model=list[VehicleResponse])
async def list_all(_=Depends(get_current_user)):
    vehicles = await list_vehicles()
    return [_to_response(v) for v in vehicles]


@router.put("/{vehicle_id}", response_model=VehicleResponse)
async def update(vehicle_id: str, data: VehicleUpdateRequest, _=Depends(get_current_user)):
    vehicle = await update_vehicle(vehicle_id, data)
    return _to_response(vehicle)


@router.delete("/{vehicle_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete(vehicle_id: str, _=Depends(require_admin)):
    await delete_vehicle(vehicle_id)

File: frontend/src/api/vehicles.ts
import client from './client'

export interface Vehicle {
  id: string
  make: string
  model: string
  category: string
  price: number
  quantity: number
}

export interface VehicleInput {
  make: string
  model: string
  category: string
  price: number
  quantity: number
}

export interface SearchFilters {
  make?: string
  category?: string
  min_price?: number
  max_price?: number
}

export async function listVehicles(): Promise<Vehicle[]> {
  const res = await client.get('/api/vehicles')
  return res.data
}

export async function searchVehicles(filters: SearchFilters): Promise<Vehicle[]> {
  const params = Object.fromEntries(
    Object.entries(filters).filter(([, v]) => v !== undefined && v !== '')
  )
  const res = await client.get('/api/vehicles/search', { params })
  return res.data
}

export async function createVehicle(data: VehicleInput): Promise<Vehicle> {
  const res = await client.post('/api/vehicles', data)
  return res.data
}

export async function updateVehicle(id: string, data: VehicleInput): Promise<Vehicle> {
  const res = await client.put(`/api/vehicles/${id}`, data)
  return res.data
}

export async function deleteVehicle(id: string): Promise<void> {
  await client.delete(`/api/vehicles/${id}`)
}

export async function purchaseVehicle(id: string): Promise<Vehicle> {
  const res = await client.post(`/api/vehicles/${id}/purchase`)
  return res.data
}

export async function restockVehicle(id: string, quantity: number): Promise<Vehicle> {
  const res = await client.post(`/api/vehicles/${id}/restock`, { quantity })
  return res.data
}